### PR TITLE
File meta data

### PR DESCRIPTION
**Pull request by poppindouble**

### CODE REVIEW

**Comment by poppindouble** [comment]
This PR is fixing issue #38

File: src/storage/preferences.rs [hunk 1]
```diff
@@ -67,11 +56,18 @@ impl DBPreferences {
 
 impl Default for DBPreferences {
     fn default() -> Self {
+        let version_str = env!("CARGO_PKG_VERSION");
+        let db_version = match LogVersion::try_from(version_str) {
+            Ok(version) => version,
+            Err(_err) => LogVersion::new(0, 0, 0),
```
Comment (blaesus) [inline src/storage/preferences.rs:62]: Should `db_version` be a preference at all? I mean, if the code can only handle `2.0.0` log format, what's the meaning of the user setting the db_version to anything else (say `3.0.0` or `1.0.0`)?

I think `db_version` might as well just be a fixed struct, directly controller by `CARGO_PKG_VERSION` and cannot be changed or specified manually by user.

**Review by blaesus — CHANGES_REQUESTED** [review]
See comments

File: src/storage/log_version.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        } else if self.major == other.major && self.minor > other.minor {
+            return Some(Ordering::Greater);
+        } else if self.major == other.major
+            && self.minor == other.minor
+            && self.revise > other.revise
+        {
+            return Some(Ordering::Greater);
+        } else {
+            return Some(Ordering::Less);
+        }
```
[the remark this answers is not in the document]
Comment (poppindouble) [reply src/storage/log_version.rs:80]: done

File: src/storage/log_reader.rs [hunk 3]
```diff
@@ -27,12 +32,16 @@ impl LogReader {
...
+    pub fn read_all(&mut self) -> KVResult<(InstructionLogIterator, usize)> {
         let mut buffer: Vec<u8> = Vec::with_capacity(self.buf_reader.capacity());
         self.buf_reader.seek(SeekFrom::Start(0))?;
         self.buf_reader.read_to_end(&mut buffer)?;
 
-        let iterator = InstructionLogIterator::from(buffer);
-        return Ok(iterator);
+        let (target_log_version, offset) = LogVersion::parse(&buffer)?;
+
+        assert!(self.log_version >= target_log_version);
```
[the remark this answers is not in the document]
Comment (poppindouble) [reply src/storage/log_reader.rs:42]: done

File: src/storage/preferences.rs (hunk 1 above)
Comment (poppindouble) [reply src/storage/preferences.rs:62]: done

File: src/storage/preferences.rs [hunk 4]
```diff
@@ -67,11 +56,18 @@ impl DBPreferences {
 
 impl Default for DBPreferences {
     fn default() -> Self {
+        let version_str = env!("CARGO_PKG_VERSION");
```
[the remark this answers is not in the document]
Comment (poppindouble) [reply src/storage/preferences.rs:59]: done

**Review by blaesus — APPROVED** [review]

**Pull request merged**